Add `sentry-dart-plugin` documentation (#7097)
getsentry/sentry-docs: src/platforms/flutter/upload-debug.mdx

---
title: Debug Symbols
sidebar_order: 2

> description: "Learn more about uploading debug symbols for both Android, iOS/macOS & Flutter Web."

Reviewer: ```suggestion
description: "Learn more about uploading debug symbols for both Android, iOS/macOS and Flutter Web."
```

Errors raised from the native layer in Flutter apps require certain debug information files to be uploaded. For example, an Android app can use `proguard` for minification and obfuscation. And when using NDK, dwarf debug files need to be uploaded. Flutter Web requires sourcemaps and iOS apps also require dwarf debug information files.

## Automatically Upload Debug Symbols

> The most straight forward way is to [use the Sentry Dart Plugin](https://github.com/getsentry/sentry-dart-plugin), which uploads debug symbols automatically for you.

Reviewer 2: ```suggestion
The easiest way to upload debug symbols is to use the [Sentry Dart Plugin](https://github.com/getsentry/sentry-dart-plugin) which will upload them automatically.
```

### Install

In your `pubspec.yaml`, add `sentry_dart_plugin` as a new dev dependency.

```yaml
dev_dependencies:
  sentry_dart_plugin: ^1.0.0
```

### Build

> The `flutter build apk`, `flutter build ios` (or _macos_) or `flutter build web` is required before executing the `sentry_dart_plugin` plugin, because the build spits out the debug symbols and source maps.

Reviewer 2: ```suggestion
Run
- `flutter build apk`
- `flutter build ios` (or `macos`), or
- `flutter build web`

before executing the `sentry_dart_plugin` plugin.

This build step outputs the debug symbols and source maps that the plugin will upload.
```

### Configure

> This tool comes with a default configuration. You can configure it to suit your needs.
> 
> Add `sentry:` configuration at the end of your `pubspec.yaml` file:

Reviewer 2: ```suggestion
This Sentry Dart Plugin comes with a default configuration.

To modify this configuration, add a `sentry:` configuration to the end of your `pubspec.yaml` file:
```

```yaml
sentry:
  upload_debug_symbols: true
  upload_source_maps: false
  upload_sources: false
  project: ...
  org: ...
  auth_token: ...
  url: ...
  wait_for_processing: false

>   log_level: error # possible values: trace, debug, info, warn, error

Reviewer: ```suggestion
  log_level: error
```

  release: ...
  web_build_path: ...
  commits: auto
  ignore_missing: true
```

#### Available Configuration Fields

`project`

> : The project's name. e.g. `sentry-flutter`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_PROJECT

Reviewer 2: ```suggestion
: The project's name, for example`sentry-flutter`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_PROJECT`.
```

`org`

> : Your organization's slug e.g. `sentry-sdks`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_ORG`.

Reviewer 2: ```suggestion
: Your organization's slug, for example`sentry-sdks`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_ORG`.
```

`auth_token`

> : The sentry auth token e.g. `<64 random characteres>`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_AUTH_TOKEN`.

Reviewer 2: ```suggestion
: The sentry auth token, which will look like`<64 random characters>`. Required. This is a `string` type. The alternative environmental variable is `SENTRY_AUTH_TOKEN`.
```

`upload_debug_symbols`

: Enables or disables the automatic upload of debug symbols. This is a boolean type with default value `true`.

`upload_source_maps`

: Enables or disables the automatic upload of source maps. This is a boolean type with default value `false`.

`upload_sources`

: Does or doesn't include the source code of native code. This is a boolean type with default value `false`.

`url`

> : The URL of your project e.g. `https<area>://mysentry.invalid/`. This is a `string` type. The alternative environmental variable is `SENTRY_URL`.

Reviewer 2: ```suggestion
: The URL of your project, for example, `https<area>://mysentry.invalid/`. This is a `string` type. The alternative environmental variable is `SENTRY_URL`.
```

`web_build_path`

: The web build folder. This is a `string` type with default value `build/web`.

`commits`

: Release commits integration. This is a `string` type with default value `auto`.

`ignore_missing`

: Ignore missing commits previously used in the release. This is a boolean type with default value `false`.

### Run

```bash
flutter packages pub run sentry_dart_plugin
```

### Troubleshooting

Refer to [Troubleshooting - Sentry Dart Plugin](/platforms/flutter/troubleshooting#sentry-dart-plugin) to resolve potential issues.

## Manually Upload Debug Symbols

### Uploading for iOS and macOS

Sentry requires a dSYM upload to symbolicate your crash logs. The symbolication process unscrambles Apple’s crash logs to reveal the function, file names, and line numbers of the crash. [Learn how to upload the dSYM files](/platforms/apple/dsym/).

If you use the `split-debug-info` and `obfuscate` flags, you need to upload the `*.dSYM` files instead of the `*.symbols` files generated by the Flutter build. The `*.dSYM` files are located in the `build` folder of your project and not the given `split-debug-info` folder.

### Uploading for Android NDK

> The NDK symbols are uploaded by the Sentry Dart Plugin as well if `upload_debug_symbols` option is enabled. As an alternative, you can [upload files manually using our CLI](/product/cli/dif/#uploading-files).

Reviewer 2: ```suggestion
The Sentry Dart Plugin will also upload NDK symbols if `upload_debug_symbols` is enabled. Alternatively, see our docs on uploading [Debug Information Files](/product/cli/dif/#uploading-files) manually with the Sentry CLI.
```

> The NDK symbols are uploaded by the Sentry Dart Plugin as well if `upload_debug_symbols` option is enabled. As an alternative, you can also use our CLI to [upload source maps for Flutter Web](https://github.com/product/cli/releases/#managing-release-artifacts).

Reviewer: https://github.com/product/cli/releases/#managing-release-artifacts is broken, which one did you mean @denrase ?

Author: This is a snippet that was there already. I assumed it was correct. 🤔 
```
## Uploading Source Maps for Flutter Web

You can also use our CLI to [upload source maps for Flutter Web](/product/cli/releases/#managing-release-artifacts).
```

Author: Must have been a CP error, sorry for the confusion.

> If you are using `sentry_flutter` version earlier than 5.1, native symbolication on Android requires a specific configuration. Refer to [Troubleshooting](/platforms/flutter/troubleshooting/#native-symbolication-on-android) for more information.

Reviewer 2: ```suggestion
If you are using a version of `sentry_flutter` earlier than 5.1, native symbolication on Android requires a specific configuration. Refer to [Troubleshooting](/platforms/flutter/troubleshooting/#native-symbolication-on-android) for more information.
```

> If you are using `sentry_flutter` version earlier than 5.1, native symbolication on Android requires a specific configuration. Refer to [Troubleshooting](/platforms/flutter/troubleshooting/) for more information.

Reviewer: Link with the `troubleshooting` fragment, so people can go directly to that section.

<Note>

Sentry's Flutter SDK doesn't currently support the `uploadNativeSymbols` flag from the [Sentry Gradle Plugin](/platforms/android/gradle/).

</Note>

### Uploading with ProGuard

> To use ProGuard with Sentry, upload the [Android Proguard/R8 mapping files](/platforms/android/proguard/) by either the recommended method of using our Gradle integration or manually by using [sentry-cli](/product/cli/dif/#proguard-mapping-upload). This is optional, and only needed if you have ProGuard (`minifyEnabled`) enabled.

Reviewer 2: ```suggestion
If you have ProGuard (`minifyEnabled`) enabled, you must upload the [Android Proguard/R8 mapping files](/platforms/android/proguard/) to see complete stack traces. You can upload these files by either using our Gradle integration (recommended) or manually by using the [Sentry CLI](/product/cli/dif/#proguard-mapping-upload).
```

### Uploading Source Maps for Flutter Web

> The Sentry Dart plugin uploads source maps as well if `upload_source_maps` is enabled. As an alternative, you can also use our CLI to [upload source maps for Flutter Web](/product/cli/releases/#managing-release-artifacts).

Reviewer 2: ```suggestion
The Sentry Dart plugin also uploads source maps if `upload_source_maps` is enabled. Alternatively, you can use the Sentry CLI to upload your source maps for Flutter Web by following the docs on [Managing Release Artifacts](/product/cli/releases/#managing-release-artifacts).
```

## Automatically Upload Debug Symbols

The most straight forward way is to [use the Sentry Dart Plugin](https://github.com/getsentry/sentry-dart-plugin), which uploads debug symbols automatically for you.

### Install

In your `pubspec.yaml`, add `sentry_dart_plugin` as a new dev dependency.

```yaml
dev_dependencies:
  sentry_dart_plugin: ^1.0.0
```

### Build

The `flutter build apk`, `flutter build ios` (or _macos_) or `flutter build web` is required before executing the `sentry_dart_plugin` plugin, because the build spits out the debug symbols and source maps.

### Configure

This tool comes with a default configuration. You can configure it to suit your needs.

Add `sentry:` configuration at the end of your `pubspec.yaml` file:

```yaml
sentry:
  upload_debug_symbols: true
  upload_source_maps: false
  upload_sources: false
  project: ...
  org: ...
  auth_token: ...
  url: ...
  wait_for_processing: false

>   log_level: error # possible values: trace, debug, info, warn, error

Reviewer: We can probably put the `possible values` in the available configuration fields.

`commits`

: Release commits integration. This is a `string` type with default value `auto`.

`ignore_missing`

: Ignore missing commits previously used in the release. This is a boolean type with default value `false`.

### Run

```bash
flutter packages pub run sentry_dart_plugin
```

### Troubleshooting

Refer to [Troubleshooting - Sentry Dart Plugin](/platforms/flutter/troubleshooting#sentry-dart-plugin) to resolve potential issues.

## Manually Upload Debug Symbols

### Uploading for iOS and macOS

Sentry requires a dSYM upload to symbolicate your crash logs. The symbolication process unscrambles Apple’s crash logs to reveal the function, file names, and line numbers of the crash. [Learn how to upload the dSYM files](/platforms/apple/dsym/).

If you use the `split-debug-info` and `obfuscate` flags, you need to upload the `*.dSYM` files instead of the `*.symbols` files generated by the Flutter build. The `*.dSYM` files are located in the `build` folder of your project and not the given `split-debug-info` folder.

### Uploading for Android NDK

The NDK symbols are uploaded by the Sentry Dart Plugin as well if `upload_debug_symbols` option is enabled.

> Check out [Uploading Files](/product/cli/dif/#uploading-files) if you want to upload manually directly with sentry-cli.

Reviewer: Let's agree on the fallback phrase for the CLI.

This one is:

> Check out [Uploading Files](/product/cli/dif/#uploading-files) if you want to upload manually directly with sentry-cli.

The other one is:

> . As an alternative, you can also use our CLI to [upload source maps for Flutter Web](/product/cli/releases/#managing-release-artifacts).

Both are fine, let's just stick to one of them, the later is nicer IMO.

Errors raised from the native layer in Flutter apps require certain debug information files to be uploaded. For example, an Android app can use `proguard` for minification and obfuscation. And when using NDK, dwarf debug files need to be uploaded. Flutter Web requires sourcemaps and iOS apps also require dwarf debug information files.

## Automatically Upload Debug Symbols

The most straight forward way is to [use the Sentry Dart Plugin](https://github.com/getsentry/sentry-dart-plugin), which uploads debug symbols automatically for you.

> ### :clipboard: Install

Reviewer: How does `:clipboard:` work? we don't use this anywhere else

Author: Ah, this is an emoji on gh. 🤦

In your `pubspec.yaml`, add `sentry_dart_plugin` as a new dev dependency.

```yaml
dev_dependencies:
  sentry_dart_plugin: ^1.0.0
```

### Build

The `flutter build apk`, `flutter build ios` (or _macos_) or `flutter build web` is required before executing the `sentry_dart_plugin` plugin, because the build spits out the debug symbols and source maps.

> ### Run

Reviewer: We have to document the configuration block before the `Run`.

https://github.com/getsentry/sentry-dart-plugin#configuration-optional
Including all the options.

We can make a table similar to https://docs.sentry.io/product/cli/dif/#upload-options

```bash
flutter packages pub run sentry_dart_plugin
```

## Manually Upload Debug Symbols

### Uploading for iOS and macOS

Sentry requires a dSYM upload to symbolicate your crash logs. The symbolication process unscrambles Apple’s crash logs to reveal the function, file names, and line numbers of the crash. [Learn how to upload the dSYM files](/platforms/apple/dsym/).

If you use the `split-debug-info` and `obfuscate` flags, you need to upload the `*.dSYM` files instead of the `*.symbols` files generated by the Flutter build. The `*.dSYM` files are located in the `build` folder of your project and not the given `split-debug-info` folder.

### Uploading for Android NDK

> The NDK symbols are uploaded by the Sentry Dart Plugin as well.

Reviewer: We can make a sentence that you have to enable the `upload_debug_symbols` option here.

A fallback back sentence should lead to `/product/cli/dif/#uploading-files` in case people wanna upload manually directly with sentry-cli.

If you are using `sentry_flutter` version earlier than 5.1, native symbolication on Android requires a specific configuration. Refer to [Troubleshooting](/platforms/flutter/troubleshooting/) for more information.

<Note>

Sentry's Flutter SDK doesn't currently support the `uploadNativeSymbols` flag from the [Sentry Gradle Plugin](/platforms/android/gradle/).

</Note>

### Uploading with ProGuard

> To use ProGuard with Sentry, upload the [Android Proguard/R8 mapping files](/platforms/android/proguard/) by either the recommended method of using our Gradle integration or manually by using [sentry-cli](/product/cli/dif/#proguard-mapping-upload). This is optional, and only needed if you have `proguard` enabled.

Reviewer: ```suggestion
To use ProGuard with Sentry, upload the [Android Proguard/R8 mapping files](/platforms/android/proguard/) by either the recommended method of using our Gradle integration or manually by using [sentry-cli](/product/cli/dif/#proguard-mapping-upload). This is optional, and only needed if you have ProGuard (`minifyEnabled`) enabled.
```

### Uploading Source Maps for Flutter Web

> The Sentry Dart plugin uploads source maps as well. You can also use our CLI to [upload source maps for Flutter Web](/product/cli/releases/#managing-release-artifacts).

Reviewer: Similar to https://github.com/getsentry/sentry-docs/pull/7097/files#r1219455786
with the `upload_source_maps` flag.
